docs/design.md

```diff
@@ -424,29 +424,39 @@ supplied to provide an upper bound on timestamps for already-committed
 data (`ε` is the maximum clock skew). As the transaction progresses, any
 data read which have timestamps greater than `t` but less than `t+ε`
 cause the transaction to abort and retry with the conflicting timestamp
-t<sub>c</sub>, where t<sub>c</sub> \> t. The maximum timestamp `t+ε` remains the same.
-Time spent retrying because of reading recently committed data has an
-upper bound of `ε`. In fact, this is further optimized: upon restarting,
-the transaction not only takes into account the timestamp of the future
-value, but the timestamp of the node at the time of the uncertain read.
-The larger of those two timestamps (typically, the latter) is used to
-bump up the read timestamp, and additionally the node is marked as
-“certain”. This means that for future reads to that node within the
-transaction, we can set `MaxTimestamp = Read Timestamp` (and hence avoid
-further uncertainty restarts). Correctness follows from the fact that we
+t<sub>c</sub>, where t<sub>c</sub> \> t. The maximum timestamp `t+ε` remains
+the same.
+
+We apply another optimization to avoid further restarts caused
+by uncertainty. Upon restarting, the transaction not only takes
+into account t<sub>c<sub>, but the timestamp of the node at the time
+of the uncertain read t<>node<>. The larger of those two timestamps:
+max(t<sub>c<sub>, t<sub>node<sub>) (more likely the latter) is used
+to bump up the read timestamp. The use of t<sub>node<sub> guarantees that the
+time spent retrying because of reading recently committed data has an
+upper bound of `ε`.
+
+Additionally, the conflicting node is marked as “certain”. This means that
+for future reads to that node within the transaction, we can set
+`MaxTimestamp = Read Timestamp`. Correctness follows from the fact that we
 know that at the time of the read, there exists no version of any key on
-that node with a higher timestamp; if we ran into one during a future
-read, that node would have happened (in absolute time) after our
-transaction started.
+that node with a higher timestamp that was updated in the past in absolute
+time. A higher timestamp is indicative that the update happened in the future
+in absolute time, and the transaction will move forward reading an older
+version of the data (at the transactions timestamp). (This seems wrong; it only
+guarantees that a transaction coordinated from this node cannot have a greater
+timestamp. You can still have transactions coordinated from other nodes with
```

Comment:
I've updated this PR. But what's missing in the discussion is the existence
of a reader cache on each node that keeps track of recently read local
values. The reader cache is used to restart transactions that contain
writes to values  in the cache with future cached time stamps. I'll update
the doc such that a discussion about the hybrid clock is above all the
discussion about transactions. Please resist reviewing for now. Thanks

On Thu, May 28, 2015, 4:55 PM Tobias Schottdorf <notifications@github.com>
wrote:

> In docs/design.md
> <https://github.com/cockroachdb/cockroach/pull/1186#discussion_r31276827>:
>
> >  know that at the time of the read, there exists no version of any key on
> > -that node with a higher timestamp; if we ran into one during a future
> > -read, that node would have happened (in absolute time) after our
> > -transaction started.
> > +that node with a higher timestamp that was updated in the past in absolute
> > +time. A higher timestamp is indicative that the update happened in the future
> > +in absolute time, and the transaction will move forward reading an older
> > +version of the data (at the transactions timestamp). (This seems wrong; it only
> > +guarantees that a transaction coordinated from this node cannot have a greater
> > +timestamp. You can still have transactions coordinated from other nodes with
>
> yes that should be correct.
>
> As for the timestamp, there's a section on the hybrid logical clock, but
> the Transaction section should probably make it clear how the Txn timestamp
> is chosen (that it's according to a local clock, and that requests sent on
> its behalf will always have that or a higher timestamp in their request
> header); feel free to update that as well if it's not there.
>
> Glad you're looking at this stuff.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1186/files#r31276827>.
>